File: python/tests/test_transport_dataset_provenance.py
"""v0.11 Task 2.5 (transport-spec §4.21): exercise the
``DATASET_PROVENANCE`` (0x18) packet on :class:`TransportWriter` +
:class:`TransportReader`.

Wire layout per §4.21::

    record_count:        uint32
    # repeated record_count times:
    timestamp_unix:      int64
    software_length:     uint16
    software:            bytes[software_length]      # UTF-8
    parameters_length:   uint16
    parameters_json:     bytes[parameters_length]    # UTF-8 JSON
    input_refs_length:   uint16
    input_refs_csv:      bytes[input_refs_length]    # UTF-8 comma-joined
    output_refs_length:  uint16
    output_refs_csv:     bytes[output_refs_length]   # UTF-8 comma-joined

All multi-byte integers are LITTLE-ENDIAN (spec §1.7).

Python parity for Java's ``TransportDatasetProvenanceTest`` (commit
``563e09c3``).

SPDX-License-Identifier: Apache-2.0
"""
from __future__ import annotations

import io
import struct
from pathlib import Path

from ttio.provenance import ProvenanceRecord
from ttio.spectral_dataset import SpectralDataset
from ttio.transport.codec import TransportReader, TransportWriter
from ttio.transport.packets import PacketType, TRANSPORT_V0_11_FEATURE


def _read_le_string(buf: bytes, offset: int, width: int) -> tuple[str, int]:
    """Read a length-prefixed UTF-8 string at ``offset``. ``width``
    selects uint16 (=2) or uint32 (=4) prefix."""
    if width == 2:
        (length,) = struct.unpack_from("<H", buf, offset)
        offset += 2
    else:
        (length,) = struct.unpack_from("<I", buf, offset)
        offset += 4
    s = buf[offset:offset + length].decode("utf-8")
    return s, offset + length


def test_write_dataset_provenance_emits_single_packet_with_record_count() -> None:
    """Writer's low-level helper emits a single 0x18 packet whose
    payload matches §4.21 exactly."""
    r1 = ProvenanceRecord(
        timestamp_unix=1700000000,
        software="TTI-O Python 1.0.0",
        parameters={"threshold": "0.5"},
        input_refs=["file:///in.raw", "file:///in2.raw"],
        output_refs=["file:///out.tio"],
    )
    r2 = ProvenanceRecord(
        timestamp_unix=1700000100,
        software="step 2",
        parameters={},
        input_refs=[],
        output_refs=["file:///final.tio"],
    )

    out = io.BytesIO()
    with TransportWriter(out) as w:
        w.write_stream_header(
            format_version="1.2",
            title="prov-test",
            isa_investigation="isa",
            features=[],
            n_datasets=0,
        )
        w.write_dataset_provenance([r1, r2])
        w.write_end_of_stream()

    r = TransportReader(io.BytesIO(out.getvalue()))
    records = r.records_for_test()
    assert len(records) == 3, (
        f"expected StreamHeader + DatasetProvenance + EndOfStream, "
        f"got {len(records)}"
    )
    assert records[1].header.packet_type == int(
        PacketType.DATASET_PROVENANCE
    )

    payload = records[1].payload
    off = 0
    (record_count,) = struct.unpack_from("<I", payload, off)
    off += 4
    assert record_count == 2

    # Record 0.
    (ts0,) = struct.unpack_from("<q", payload, off)
    off += 8
    assert ts0 == 1700000000
    software0, off = _read_le_string(payload, off, 2)
    assert software0 == "TTI-O Python 1.0.0"
    params0, off = _read_le_string(payload, off, 2)
    assert "threshold" in params0
    assert "0.5" in params0
    inputs0, off = _read_le_string(payload, off, 2)
    assert inputs0 == "file:///in.raw,file:///in2.raw"
    outputs0, off = _read_le_string(payload, off, 2)
    assert outputs0 == "file:///out.tio"

    # Record 1 (empty params/inputs).
    (ts1,) = struct.unpack_from("<q", payload, off)
    off += 8
    assert ts1 == 1700000100
    software1, off = _read_le_string(payload, off, 2)
    assert software1 == "step 2"
    params1, off = _read_le_string(payload, off, 2)
    assert params1 == "{}", (
        f"empty parameters must render as '{{}}', got {params1!r}"
    )
    inputs1, off = _read_le_string(payload, off, 2)
    assert inputs1 == "", (
        f"empty input_refs must render as '', got {inputs1!r}"
    )
    outputs1, off = _read_le_string(payload, off, 2)
    assert outputs1 == "file:///final.tio"
    assert off == len(payload), (
        "payload must contain only the 2 records, no trailing bytes"
    )


def test_write_dataset_provenance_zero_records_emits_no_packet() -> None:
    """An empty record list should not emit a 0x18 packet at all
    (spec §5.4 "zero or more")."""
    out = io.BytesIO()
    with TransportWriter(out) as w:
        w.write_stream_header(
            format_version="1.2",
            title="empty",
            isa_investigation="",
            features=[],
            n_datasets=0,
        )
        w.write_dataset_provenance([])
        w.write_end_of_stream()

    r = TransportReader(io.BytesIO(out.getvalue()))
    for rec in r.records_for_test():
        assert rec.header.packet_type != int(PacketType.DATASET_PROVENANCE), (
            "empty record list must not emit DATASET_PROVENANCE"
        )


def test_write_dataset_emits_dataset_provenance_when_present(
    tmp_path: Path,
) -> None:
    """``write_dataset`` on a ``.tio`` carrying provenance emits
    exactly one DATASET_PROVENANCE packet (in the v0.11 prelude, after
    ENCRYPTION_ALGORITHM, before reference groups per §5.4) AND flips
    the v0.11 feature flag."""
    src = tmp_path / "prov.tio"
    _build_provenance_only(src)
    tis = tmp_path / "prov.tis"

    with SpectralDataset.open(src) as ds:
        assert ds.provenance(), (
            "fixture precondition: dataset must carry provenance"
        )
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(tis.read_bytes()))
    records = r.records_for_test()
    # StreamHeader features must include transport_v0_11.
    from ttio.transport.codec import _decode_stream_header
    sh = _decode_stream_header(records[0].payload)
    assert TRANSPORT_V0_11_FEATURE in sh["features"], (
        "StreamHeader must carry transport_v0_11 feature flag"
    )

    prov_count = sum(
        1 for rec in records
        if rec.header.packet_type == int(PacketType.DATASET_PROVENANCE)
    )
    assert prov_count == 1, (
        f"write_dataset on provenance-bearing .tio must emit exactly "
        f"one DATASET_PROVENANCE packet, got {prov_count}"
    )


def test_write_dataset_no_packet_when_provenance_empty(
    tmp_path: Path,
) -> None:
    """``write_dataset`` on a ``.tio`` with NO provenance emits NO
    0x18 packet."""
    src = tmp_path / "plain.tio"
    SpectralDataset.write_minimal(
        src,
        title="plain",
        isa_investigation_id="",
        runs={},
    )
    tis = tmp_path / "plain.tis"
    with SpectralDataset.open(src) as ds:
        assert not ds.provenance(), (
            "fixture precondition: dataset must carry no provenance"
        )
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(tis.read_bytes()))
    for rec in r.records_for_test():
        assert rec.header.packet_type != int(PacketType.DATASET_PROVENANCE), (
            "empty-provenance dataset must not emit DATASET_PROVENANCE"
        )


def test_dataset_provenance_round_trips_three_records(tmp_path: Path) -> None:
    """End-to-end round-trip preserves record count + per-record
    fields + record order."""
    src = tmp_path / "src.tio"
    _build_provenance_only(src)
    tis = tmp_path / "src.tis"
    rt = tmp_path / "rt.tio"

    with SpectralDataset.open(src) as ds:
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    with TransportReader(io.BytesIO(tis.read_bytes())) as r:
        materialised = r.read_to_dataset(output_path=rt)
        materialised.close()

    with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
        prov_a = a.provenance()
        prov_b = b.provenance()
        assert len(prov_a) == len(prov_b), (
            f"record count mismatch: src={len(prov_a)}, rt={len(prov_b)}"
        )
        for i, (ra, rb) in enumerate(zip(prov_a, prov_b)):
            assert ra.timestamp_unix == rb.timestamp_unix, (
                f"timestamp_unix mismatch at record {i}: "
                f"{ra.timestamp_unix} vs {rb.timestamp_unix}"
            )
            assert ra.software == rb.software, (
                f"software mismatch at record {i}"
            )
            assert list(ra.input_refs) == list(rb.input_refs), (
                f"input_refs mismatch at record {i}"
            )
            assert list(ra.output_refs) == list(rb.output_refs), (
                f"output_refs mismatch at record {i}"
            )
            # parameters: compare as dicts (the on-disk JSON form may
            # reorder keys, but the dict contents must match).
            assert dict(ra.parameters) == dict(rb.parameters), (
                f"parameters mismatch at record {i}"
            )


def _build_provenance_only(target: Path) -> None:
    """Build a minimal ``.tio`` carrying 3 provenance records
    (1 with non-empty params + multi-element refs, 1 with all-empty
    fields, 1 with single-element refs). Exercises every empty/non-
    empty combination of the v0.11 wire layout."""
    r1 = ProvenanceRecord(
        timestamp_unix=1700000000,
        software="TTI-O Python 1.0.0",
        parameters={"mode": "strict", "threshold": "0.5"},
        input_refs=["file:///in.raw", "file:///in2.raw"],
        output_refs=["file:///out.tio"],
    )
    r2 = ProvenanceRecord(
        timestamp_unix=1700000100,
        software="downstream step",
        parameters={},
        input_refs=[],
        output_refs=[],
    )
    r3 = ProvenanceRecord(
        timestamp_unix=1700000200,
        software="final step",
        parameters={"k": "v"},
        input_refs=["file:///r2-out.tio"],
        output_refs=["file:///final.tio"],
    )
    SpectralDataset.write_minimal(
        target,
        title="provenance_only",
        isa_investigation_id="",
        runs={},
        provenance=[r1, r2, r3],
    )

File: objc/Tests/TestTransportDatasetProvenance.m
/*
 * TestTransportDatasetProvenance.m — Task 3.5 of transport-spec v0.11.
 *
 * Exercises DATASET_PROVENANCE (0x18) writer + reader. Wire layout
 * per transport-spec §4.21: single packet, `uint32 record_count`
 * prefix + N records, per record:
 *   int64  timestamp_unix
 *   uint16 software_length      + UTF-8 bytes
 *   uint16 parameters_length    + UTF-8 JSON
 *   uint16 input_refs_length    + UTF-8 CSV
 *   uint16 output_refs_length   + UTF-8 CSV
 * All LE.
 *
 * Cross-language parity:
 *   Java TransportDatasetProvenanceTest (commit 563e09c3)
 *   Python tests/test_transport_dataset_provenance.py (commit 434d45a6)
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"

#import "Transport/TTIOTransportPacket.h"
#import "Transport/TTIOTransportReader.h"
#import "Transport/TTIOTransportReader+Internal.h"
#import "Transport/TTIOTransportWriter.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOProvenanceRecord.h"
#include <unistd.h>

static NSString *makeTempPathP(NSString *suffix)
{
    NSString *base = [NSString stringWithFormat:@"ttio_tr_prov_%d_%@.tio",
                       (int)getpid(), suffix];
    return [NSTemporaryDirectory()
        stringByAppendingPathComponent:base];
}

// ---------- LE readers ----------

static uint16_t leU16P(const uint8_t *b)
{
    return (uint16_t)((uint32_t)b[0] | ((uint32_t)b[1] << 8));
}

static uint32_t leU32P(const uint8_t *b)
{
    return (uint32_t)b[0]
         | ((uint32_t)b[1] << 8)
         | ((uint32_t)b[2] << 16)
         | ((uint32_t)b[3] << 24);
}

static int64_t leI64P(const uint8_t *b)
{
    uint64_t lo = (uint64_t)leU32P(b);
    uint64_t hi = (uint64_t)leU32P(b + 4);
    return (int64_t)(lo | (hi << 32));
}

static NSString *readLEStrP(const uint8_t *bytes, NSUInteger *off, uint16_t *outLen)
{
    uint16_t len = leU16P(&bytes[*off]); *off += 2;
    if (outLen) *outLen = len;
    NSString *s = [[NSString alloc] initWithBytes:&bytes[*off]
                                            length:len
                                          encoding:NSUTF8StringEncoding] ?: @"";
    *off += len;
    return s;
}

// ---------- writer low-level: zero records is a no-op ----------

static void testWriteDatasetProvenanceZeroRecordsEmitsNoPacket(void)
{
    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"prov-empty"
                               isaInvestigation:@""
                                       features:@[]
                                      nDatasets:0
                                          error:&err],
         "3.5 zero: StreamHeader");
    PASS([w writeDatasetProvenance:@[] error:&err],
         "3.5 zero: writeDatasetProvenance:[] succeeded");
    PASS([w writeEndOfStreamWithError:&err], "3.5 zero: EndOfStream");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records.count == 2,
         "3.5 zero: 0 records -> only StreamHeader + EndOfStream");
    for (TTIOTransportPacketRecord *rec in records) {
        PASS(rec.header.packetType
                != TTIOTransportPacketDatasetProvenance,
             "3.5 zero: no 0x18 packet emitted");
    }
}

// ---------- writer low-level: single record round-trips ----------

static TTIOProvenanceRecord *makeRecord(int64_t ts,
                                          NSString *software,
                                          NSDictionary *params,
                                          NSArray *inputs,
                                          NSArray *outputs)
{
    return [[TTIOProvenanceRecord alloc]
                initWithInputRefs:inputs
                         software:software
                       parameters:params
                       outputRefs:outputs
                    timestampUnix:ts];
}

static void testWriteDatasetProvenanceSingleRecord(void)
{
    TTIOProvenanceRecord *r1 = makeRecord(
        1700000000LL,
        @"TTI-O ObjC 1.0.0",
        @{@"threshold": @"0.5"},
        @[@"file:///in.raw", @"file:///in2.raw"],
        @[@"file:///out.tio"]);

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"prov-one"
                               isaInvestigation:@""
                                       features:@[]
                                      nDatasets:0
                                          error:&err],
         "3.5 one: StreamHeader");
    PASS([w writeDatasetProvenance:@[r1] error:&err],
         "3.5 one: writeDatasetProvenance emitted");
    PASS([w writeEndOfStreamWithError:&err], "3.5 one: EndOfStream");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records.count == 3,
         "3.5 one: StreamHeader + DatasetProvenance + EndOfStream");
    PASS(records[1].header.packetType
            == TTIOTransportPacketDatasetProvenance,
         "3.5 one: middle packet is DATASET_PROVENANCE (0x18)");

    NSData *payload = records[1].payload;
    const uint8_t *bytes = payload.bytes;
    NSUInteger off = 0;
    uint32_t count = leU32P(&bytes[off]); off += 4;
    PASS(count == 1, "3.5 one: record_count == 1");
    int64_t ts = leI64P(&bytes[off]); off += 8;
    PASS(ts == 1700000000LL, "3.5 one: timestamp_unix LE round-trips");
    uint16_t softLen = 0;
    NSString *software = readLEStrP(bytes, &off, &softLen);
    PASS([software isEqualToString:@"TTI-O ObjC 1.0.0"],
         "3.5 one: software string decoded");
    NSString *paramsJson = readLEStrP(bytes, &off, NULL);
    PASS([paramsJson isEqualToString:@"{\"threshold\":\"0.5\"}"],
         "3.5 one: parameters_json sorted-keys + no spaces");
    NSString *inputsCsv = readLEStrP(bytes, &off, NULL);
    PASS([inputsCsv isEqualToString:@"file:///in.raw,file:///in2.raw"],
         "3.5 one: input_refs comma-joined");
    NSString *outputsCsv = readLEStrP(bytes, &off, NULL);
    PASS([outputsCsv isEqualToString:@"file:///out.tio"],
         "3.5 one: output_refs comma-joined");
    PASS(off == payload.length,
         "3.5 one: no trailing bytes in payload");
}

// ---------- writer low-level: 3 records preserve order ----------

static void testWriteDatasetProvenanceThreeRecordsPreservesOrder(void)
{
    TTIOProvenanceRecord *r1 = makeRecord(
        1700000000LL, @"step A", @{@"a": @"1"},
        @[@"file:///in1"], @[@"file:///out1"]);
    TTIOProvenanceRecord *r2 = makeRecord(
        1700000100LL, @"step B", @{},
        @[], @[]);
    TTIOProvenanceRecord *r3 = makeRecord(
        1700000200LL, @"step C", @{@"k": @"v"},
        @[@"file:///x", @"file:///y"], @[@"file:///z"]);

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"prov-three"
                               isaInvestigation:@""
                                       features:@[]
                                      nDatasets:0
                                          error:&err],
         "3.5 three: StreamHeader");
    NSArray *threeRecs = @[r1, r2, r3];
    PASS([w writeDatasetProvenance:threeRecs error:&err],
         "3.5 three: writeDatasetProvenance(3 records)");
    PASS([w writeEndOfStreamWithError:&err], "3.5 three: EndOfStream");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records[1].header.packetType
            == TTIOTransportPacketDatasetProvenance,
         "3.5 three: middle packet is DATASET_PROVENANCE");

    NSData *payload = records[1].payload;
    const uint8_t *bytes = payload.bytes;
    NSUInteger off = 0;
    uint32_t count = leU32P(&bytes[off]); off += 4;
    PASS(count == 3, "3.5 three: record_count == 3");

    // Record 0
    int64_t ts0 = leI64P(&bytes[off]); off += 8;
    PASS(ts0 == 1700000000LL, "3.5 three: record 0 timestamp");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"step A"],
         "3.5 three: record 0 software");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"{\"a\":\"1\"}"],
         "3.5 three: record 0 params");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"file:///in1"],
         "3.5 three: record 0 inputs");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"file:///out1"],
         "3.5 three: record 0 outputs");

    // Record 1 — empty params + empty inputs + empty outputs
    int64_t ts1 = leI64P(&bytes[off]); off += 8;
    PASS(ts1 == 1700000100LL, "3.5 three: record 1 timestamp");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"step B"],
         "3.5 three: record 1 software");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"{}"],
         "3.5 three: empty params rendered as '{}'");
    uint16_t inLen1 = 0;
    NSString *in1 = readLEStrP(bytes, &off, &inLen1);
    PASS([in1 isEqualToString:@""] && inLen1 == 0,
         "3.5 three: empty input_refs encoded as zero-length string");
    uint16_t outLen1 = 0;
    NSString *out1 = readLEStrP(bytes, &off, &outLen1);
    PASS([out1 isEqualToString:@""] && outLen1 == 0,
         "3.5 three: empty output_refs encoded as zero-length string");

    // Record 2
    int64_t ts2 = leI64P(&bytes[off]); off += 8;
    PASS(ts2 == 1700000200LL, "3.5 three: record 2 timestamp");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"step C"],
         "3.5 three: record 2 software");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"{\"k\":\"v\"}"],
         "3.5 three: record 2 params");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"file:///x,file:///y"],
         "3.5 three: record 2 inputs comma-joined");
    PASS([readLEStrP(bytes, &off, NULL) isEqualToString:@"file:///z"],
         "3.5 three: record 2 outputs");

    PASS(off == payload.length,
         "3.5 three: payload size accounted for by 3 records");
}

// ---------- writeDataset emits 0x18 when records present ----------

static NSString *buildProvenanceOnlyTio(NSString *base, NSError **err)
{
    NSString *path = makeTempPathP(base);
    unlink([path fileSystemRepresentation]);
    TTIOProvenanceRecord *r1 = makeRecord(
        1700000000LL, @"TTI-O ObjC 1.0.0", @{@"mode": @"strict"},
        @[@"file:///in.raw"], @[@"file:///out.tio"]);
    TTIOProvenanceRecord *r2 = makeRecord(
        1700000100LL, @"downstream step", @{},
        @[], @[@"file:///final.tio"]);
    BOOL ok = [TTIOSpectralDataset writeMinimalToPath:path
                                                  title:@"provenance_only"
                                     isaInvestigationId:@""
                                                 msRuns:@{}
                                        identifications:nil
                                        quantifications:nil
                                      provenanceRecords:@[r1, r2]
                                                  error:err];
    if (!ok) return nil;
    return path;
}

static void testWriteDatasetEmitsDatasetProvenanceWhenPresent(void)
{
    NSError *err = nil;
    NSString *src = buildProvenanceOnlyTio(@"src_prov", &err);
    PASS(src != nil && err == nil, "3.5 wd: built provenance-only fixture");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil, "3.5 wd: opened fixture dataset");
    PASS(ds.provenanceRecords.count == 2,
         "3.5 wd: fixture has 2 provenance records");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.5 wd: writeDataset on provenance-bearing .tio");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records != nil, "3.5 wd: reader parsed stream");

    NSString *sh = [[NSString alloc] initWithData:records[0].payload
                                          encoding:NSUTF8StringEncoding];
    PASS(sh != nil && [sh containsString:@"transport_v0_11"],
         "3.5 wd: StreamHeader carries transport_v0_11 feature flag");

    int provCount = 0;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType
                == TTIOTransportPacketDatasetProvenance) {
            provCount++;
        }
    }
    PASS(provCount == 1,
         "3.5 wd: exactly one DATASET_PROVENANCE packet emitted");

    [ds closeFile];
    unlink([src fileSystemRepresentation]);
}

// ---------- writeDataset on no-prov .tio emits no 0x18 ----------

static void testWriteDatasetNoPacketWhenProvenanceEmpty(void)
{
    NSError *err = nil;
    NSString *src = makeTempPathP(@"plain_noprov");
    unlink([src fileSystemRepresentation]);
    BOOL ok = [TTIOSpectralDataset writeMinimalToPath:src
                                                  title:@"plain"
                                     isaInvestigationId:@""
                                                 msRuns:@{}
                                        identifications:nil
                                        quantifications:nil
                                      provenanceRecords:nil
                                                  error:&err];
    PASS(ok, "3.5 noprov: minimal .tio created");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil && ds.provenanceRecords.count == 0,
         "3.5 noprov: dataset has no provenance");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.5 noprov: writeDataset on no-prov .tio");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    BOOL sawProv = NO;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType
                == TTIOTransportPacketDatasetProvenance) {
            sawProv = YES;
            break;
        }
    }
    PASS(!sawProv,
         "3.5 noprov: no-prov dataset emits no 0x18 packet");

    [ds closeFile];
    unlink([src fileSystemRepresentation]);
}

// ---------- end-to-end round-trip via materialize ----------

static void testDatasetProvenanceRoundTripsViaMaterialize(void)
{
    NSError *err = nil;
    NSString *src = buildProvenanceOnlyTio(@"src_rt", &err);
    PASS(src != nil, "3.5 rt: built source");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil, "3.5 rt: opened source");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err], "3.5 rt: writeDataset");

    NSString *rt = makeTempPathP(@"rt");
    unlink([rt fileSystemRepresentation]);
    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    PASS([r writeTtioToPath:rt error:&err] && err == nil,
         "3.5 rt: writeTtioToPath materialised");

    TTIOSpectralDataset *back =
        [TTIOSpectralDataset readFromFilePath:rt error:&err];
    PASS(back != nil, "3.5 rt: re-opened round-trip .tio");
    PASS(back.provenanceRecords.count == ds.provenanceRecords.count,
         "3.5 rt: round-trip record count matches source");

    NSArray<TTIOProvenanceRecord *> *provA = ds.provenanceRecords;
    NSArray<TTIOProvenanceRecord *> *provB = back.provenanceRecords;
    for (NSUInteger i = 0; i < MIN(provA.count, provB.count); i++) {
        TTIOProvenanceRecord *a = provA[i];
        TTIOProvenanceRecord *b = provB[i];
        PASS(a.timestampUnix == b.timestampUnix,
             "3.5 rt: timestamp matches");
        PASS([a.software isEqualToString:b.software ?: @""],
             "3.5 rt: software matches");
        PASS([a.inputRefs isEqualToArray:(b.inputRefs ?: @[])],
             "3.5 rt: inputRefs matches");
        PASS([a.outputRefs isEqualToArray:(b.outputRefs ?: @[])],
             "3.5 rt: outputRefs matches");
    }

    [ds closeFile];
    [back closeFile];
    unlink([src fileSystemRepresentation]);
    unlink([rt fileSystemRepresentation]);
}

void testTransportDatasetProvenance(void);
void testTransportDatasetProvenance(void)
{
    testWriteDatasetProvenanceZeroRecordsEmitsNoPacket();
    testWriteDatasetProvenanceSingleRecord();
    testWriteDatasetProvenanceThreeRecordsPreservesOrder();
    testWriteDatasetEmitsDatasetProvenanceWhenPresent();
    testWriteDatasetNoPacketWhenProvenanceEmpty();
    testDatasetProvenanceRoundTripsViaMaterialize();
}
